design/one-pager-host-aware-stack-manager.md

```diff
@@ -0,0 +1,211 @@
+that will be deployed into host cluster. This way, we can explicitly declare permissions of stack manager on Host
+Cluster.
+
+#### Stack Installation
+
+In case of a stack installation, Stack Manager should:
+
+1. Create install-job on Host Cluster and read logs of install pod from Host.
+2. Deploy `Stack`, `CRDs` and `RBAC` resources on Tenant Instance
+3. Wait until a service account token secret is created for the stack to be installed.
+4. Copy service account token secret into Host Cluster.
+5. Deploy stack controller deployment by [configuring against Tenant Instance API Server](#configuring-a-pod-against-tenant-kubernetes-api-server).
+
+
+##### Mapping Resources to Host
+
+While creating resources (token secrets or controller deployments) on Host Cluster, Stack Manager will map objects
+(jobs, deployments, secrets ...) on multiple namespaces of Tenant Instance into a single namespace on Host Cluster
+by prepending `<namespace>.` to object name. For example:
+
+ | Tenant Instance | Host Cluster |
+ | :---: | :---: |
+ | Namespace / Name | Namespace / Name |
+ | crossplane-system / stack-gcp | tenant-n-controllers / crossplane-system.stack-gcp |
+ | dev / stack-wordpress | tenant-n-controllers / dev.stack-wordpress |
+
+This mapping will cause a length limit to resource names as follows:
+
+```
+len(namespace) + len(resource-name) + 1 <= 253
+```
+
+Right now, this will effect:
+- Install job name
+- Stack controller deployment name
+- Stack controller service account name
+
+However, considering the following, this should not cause a new limitation for Crossplane:
+
+- There are already labels holding some resource names (e.g. `core.crossplane.io/parent-name: stack-gcp`) and Kubernetes
+ Labels are [limited to 63 chars](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set)
+- Kubernetes namespaces are limited to 63 chars
+ 
+#### Stack Uninstallation
+
+Stack manager can no longer rely on owner references for deletion of stack install jobs and controller deployments
+since stack related custom resources and jobs/deployments live in different Kubernetes API Servers. Label based cleanup
+will be implemented as a solution.
+
+One extra artifact that needs to deleted per Stack uninstallation is the token secret created on Host Cluster. This will
+be achieved by setting owner of that secret as Stack Controller Deployment which also lives same Cluster/Namespace.
+
+### Packaging Changes
+
+We need to deploy crossplane types and roles (CRDs and RBACs) separately from Crossplane and Stack Manager pods. This
+could be realized by introducing some new helm parameters to existing Crossplane helm chart, however,
+[`helm 3` does not support templating in CRDs directory](https://github.com/helm/helm/issues/7083) which blocks
+conditional installation.
+
+We will introduce following two helm (helm 3) charts:
```

Comment:
does this change the user experience for installing Crossplane? i.e. do additional steps to install multiple helm charts get added to https://crossplane.io/docs/v0.7/install-crossplane.html?

Reply:
No, user experience for regular installation will not be effected.
We will keep existing helm chart (cluster/charts/crossplane) there and add two new helm charts.   In order not to keep/maintain same manifests (e.g. crossplane deployment) at different places, we will (automatically) build crossplane chart using the files living inside the new charts (e.g. crossplane-controllers/templates/deployment.yaml)